File: Java/Item.java

public abstract class Item {


    // variables

    private String name ;

    private double price ;

    private String description ;

    private int stock ;

    private int id ;

    
    public Item(String name,double price , String description , int stock , int id) throws IllegalCharacterException {
    	
    	
    	if(name.contains(";") || description.contains(";")) throw new IllegalCharacterException(); 
    	
    	
        this.name = name ;
        this.price = price ;
        this.description = description ;
        this.stock = stock ;
        this.id = id ;
        
    }


    // methods

    
    public String getBasicInfo() {
        
        return this.name + ";" +  Double.toString(this.price) + ";" + this.description + ";" + Integer.toString(this.stock) + ";" +Integer.toString(this.id) ;
        
    }
    
    
    public abstract String getDetails();
    
    
    @Override
    public String toString() {
        
        String[] splited = getBasicInfo().split(";") ;
        
        return "Item name: "+ splited[0] + ", Price: " + splited[1] + ", Description: " + splited[2] + ", Stock: " + splited[3] + ", ID: " + splited[4] + getDetails() ;
         
    }
    
    
//--------------------------------------Setters-----------------------------------------------
    
    
    
    public void setName(String n) throws IllegalCharacterException {          //Name 
        
    	if (n.contains(";"))
    		throw new IllegalCharacterException();
    	
        this.name = n ;
         
    }
    
    public void setPrice(double p){         //Price
        
        this.price = p ;
        
    }
    
    public void setDescription(String d) throws IllegalCharacterException {   //Description
        
    	if (d.contains(";"))
    		throw new IllegalCharacterException();
    	
        this.description = d ;
            
    }
    
    public void setStock(int s){             //Stock
        
        this.stock = s ; 
        
    }

    
    
    
    
//---------------------------------------Getters------------------------------------------------
    
    
    
    public String getName() {            //Name
        
        return this.name;
        
    }
    
    public double getPrice() {  		//Price
        
        return this.price ;
        
    }
    
    public int getStock() {				//stock
        
        return this.stock ;
        
    }
    
    public int getId() {				//Id
        
        return this.id ;
        
    }

    
    
}

File: C++/Item.cpp
#include "Item.h"

Item::Item(string n, double p, string d, int s, int i): name(n), price(p), description(d), stock(s), id(i) {

    if (n.find(";") != string::npos || d.find(";") != string::npos)
        throw IllegalCharacterException();

}

string Item::getBasicInfo() {
    ostringstream basicInfoStream;
    basicInfoStream << setprecision(2) << fixed <<this->name << ";" << this->price << ";" << this->description << ";" << this->stock << ";" << this->id;

    return basicInfoStream.str();

}

string Item::toString() {

    string data[5];


    string initial = this->getBasicInfo();
    stringstream input_sstream(initial);

    for (int i = 0; i < 5; ++i)
        getline(input_sstream, data[i], ';');

    return "Item name: " + data[0] + ", Price: " + data[1] + ", Description: " + data[2] + ", Stock: " + data[3] + ", ID: " + data[4] + this->getDetails();

}

void Item::setName(string n) {

    if (n.find(";") != string::npos)
        throw IllegalCharacterException();

    this->name = n;
}

void Item::setPrice(double p) {
    this->price = p;
}

void Item::setDescription(string d) {

    if (d.find(";") != string::npos)
        throw IllegalCharacterException();

    this->description = d;
}

void Item::setStock(int s) {
    this->stock = s;
}

string Item::getName() {
    return this->name;
}

double Item::getPrice() {
    return this->price;
}

int Item::getStock() {
    return this->stock;
}

int Item::getId() {
    return this->id;
}
